### ISSUE TITLE
Querying a 63GB file took a long time

### ISSUE THREAD

**Issue by Procrat** [BurntSushi/fst#20]
I'm using this library for querying a 63GB FST of about 7 billion strings of length 9. With more than enough memory available too completely load/cache the FST, it still took 16 minutes to only query about 15 thousand strings. Then I first copied the files to a tmpfs before doing the exact same thing and I was able to query more than 300 million strings in only 36 minutes.

Do you have any idea what might be the problem in the first case or how I would go about bisecting the issue?

**Comment by BurntSushi** [comment]
What query did you run?

**Comment by Procrat** [comment]
The queries were also all strings of length 9. (They're amino acid sequences, in case you were wondering.)

But as I'm writing this, I'm starting to guess that it's probably because it's not all cached in memory at once, but part by part read from disk, which probably makes it that slow. I also just saw the part on requiring fast random access in your blog post. So I guess it's best to forget about the memory mapping and change the code so it's completely loaded in memory then? (Feel free to close the issue then.)

**Comment by BurntSushi** [comment]
TL;DR - Yes, you're right.

Sorry, I read your comment just as I was leaving work and didn't fully absorb it. If you're doing 15,000 simple key lookups on that big of an FST and it *isn't* in memory, then it's probably spending most of its time blocked on IO. One way to confirm this would be to look at CPU and disk utilization while you're running your queries. If everything is in memory, then running 15,000 queries should probably be at 100% CPU utilization. If the FST needs to be read from disk, then yeah, it's going to be a bit slower. This also very strongly depends on what kind of disk you're using. If you're, for example, running this on an EC2 EBS disk with a low number of iops, then it's probably going to be super slow. But if you're running it directly on an SSD, at least in my experience, it's not that much slower than RAM. The reason here is that with a high latency disk (e.g., spinning rust or cheap EBS), you're paying exorbitantly for those random seeks.

Moving it to a `tmpfs` and witnessing increased throughput makes complete sense---because it's already in RAM.

Another way to get it in RAM is to put it in the file cache. e.g., `cat procrat.fst > /dev/null`. Subsequent queries should behave as if it were all in RAM.

> So I guess it's best to forget about the memory mapping and change the code so it's completely loaded in memory then?

Yes, this should also work. Note though that you'll probably pay a hefty startup cost for loading that much data on to the heap. Using `tmpfs` and memory maps means the startup cost is `0`. Up to you and your particular circumstances whether that's acceptable or not though. :-)

**Comment by BurntSushi** [comment]
> I'm using this library for querying an 63GB FST of about 7 billion strings of length 9

This stat is somewhat surprising to me. `7,000,000,000 * 9 == ~63GB`, which means you're getting virtually no compression. Wow.

**Comment by Procrat** [comment]
You're right, the CPU utilization was very low now that I look again. I'll ask what kind of disks are being used.

Thanks for all the information and that trick to put it in the file cache. :smile: I don't think loading it on the heap will be a lot slower than copying it to a `tmpfs`, right? Then I'll go for that option.

Your calculation seems right. A gzipped TSV file of the data with three fields (an index number, the sequence made up of 9 times 20 possible capitals and the integer it maps onto) sorted by sequence is 60 GB, so you seem to have about the same order of compression. If you want some more details on it, I'd be happy to provide them!

**Comment by BurntSushi** [comment]
> I don't think loading it on the heap will be a lot slower than copying it to a tmpfs, right?

I think the idea is that you can copy it to `tmpfs` *once* and then query it to your heart's content. But if you need to spin up a new process to launch a query, then it would have to load the FST into memory each time. That might be OK, depending on your use patterns. Or perhaps it's OK, but development iteration time might drop. It really depends on how your stuff is setup and what you care about. Just a suggestion. :-)

> Your calculation seems right. A gzipped TSV file of the data with three fields (an index number, the sequence made up of 9 times 20 possible capitals and the integer it maps onto) sorted by sequence is 60 GB, so you seem to have about the same order of compression. If you want some more details on it, I'd be happy to provide them!

Interesting. The key is that my calculation was for *uncompressed* data. If your data has 7 billion strings each of length 9, then storing them naively in a plain text file (one per line) is going to be about 63GB. The strange thing is that your FST is about the same size, which suggests you aren't getting any compression, which is weird. Some things to consider:

1. Your gzipped file might contain more data, e.g., the index number and integer for the map. Depending on how those numbers are represented (and their range), they could easily take up more space than the sequences themselves.
2. Is your FST a map? i.e., Does it map sequences to integers or is it just a set? Making it a map can cause compression to suffer, but the figures I'm seeing still look weird.

**Comment by bmesuere** [comment]
@BurntSushi (I'm involved in the same research project as @Procrat) It's indeed a map, mapping the strings to integers. 

Earlier experiments using strings with length between 5 and 50 gave the same results results:
* 211,885,500 strings (keys) with an estimated average length of ~13 and an integer value between 0-1.5M
* 4.55 GB uncompressed
* 1.85 GB gzip'ed
* 3.22 GB FST

**Comment by BurntSushi** [comment]
@bmesuere Interesting, thanks! A map will indeed get worse compression, but it looks like you're getting some, so that's good.

Also, you guys have made the biggest FST I've heard of. :-)

**Comment by Procrat** [comment]
Do you think using `registry_minimal.rs` might give us a better compression? I read in your blog post that using 10,000 slots in the registry gives an approximately minimal FST, but maybe that number doesn't scale that well? I think we'll try it out either way.

Thanks for all the help so far, by the way!

**Comment by BurntSushi** [comment]
@Procrat It would be at least interesting as an experiment, but you'd need quite a bit of memory to do it! It will also slow down construction time quite a bit. It's possible both of those things might make it infeasible to at the scale you're talking about, but I honestly just don't know so I would very much welcome an experiment here.
